### PR TITLE
Fix array type missing from query schema selector

### PR DESCRIPTION
**Pull request by melohagan**
## Description
This bug has been around for a while and a user wasn't able to display nested MongoDB documents, so I wanted to have a stab at it. 

This will allow you to use the default schema assigned to nested array objects, so that you can use that data in a nested table block or repeater block. For example accounts with nested product docs, etc.
The bindings also appear in the bindings drawer.

This isn't complete support for JSON/Array type columns in queries and SQL, but it allows you to at least use your data even if you can't edit the nested schemas in the UI.  

The 'Array' type will now appear in the *Schema* tab of the Query Viewer, but you cannot change it (read only).

## Addresses
- https://github.com/Budibase/budibase/issues/3382
- https://github.com/Budibase/budibase/issues/8389

## Screenshots

**BEFORE**
BEFORE_MONGODB_QUERY

**AFTER**

*Read query with nested data*
Read query

*Schema showing 'Array' type*
Screenshot 2024-01-13 at 20 13 10

*Table block using nested data example*
JSON array from query

*Card block example, using the bindings for Title, Description*
cards block example

*Bindings displaying in the drawer*
bindings from query json array

*Much more efficient than making a second nested query for the same data. Now only one POST is done by the outer data provider*
only one post query

*If the array has primitives and not nested objects, it will still work:*
non object array

### CODE REVIEW

File: packages/frontend-core/src/utils/json.js
Comment (samwho) [inline packages/frontend-core/src/utils/json.js:130]: This feels incomplete. For example, `isObject(new Map())` would return true.

**Review by samwho — COMMENTED** [review]
Have given this a once-over and nothing jumps out at me, but I'd recommend getting a review from someone more familiar with the frontend aspects of the code.

File: packages/frontend-core/src/utils/json.js
Comment (samwho) [reply packages/frontend-core/src/utils/json.js:130]: I took a look at some other implementations of `isObject` on StackOverflow and NPM and there doesn't seem to be any canonical implementation that people agree on (the `is-object` npm package has the same `new Map()` problem).

What's the goal of the function? Maybe it's fine that it's incomplete?

Comment (adrinr) [reply packages/frontend-core/src/utils/json.js:130]: If passing an empty object returns false, it seems like the function name is wrong. Could it be more verbose? What are you actually trying to get from it?

File: packages/client/src/components/app/forms/Form.svelte
Comment (adrinr) [inline packages/client/src/components/app/forms/Form.svelte:67]: Could we not be more explicit? Could we exclude the specific types, and can we use the typed ones?

File: packages/frontend-core/src/utils/json.js
Comment (melohagan) [reply packages/frontend-core/src/utils/json.js:130]: I think it probably is, specifically it's testing a a property of an API JSON response body is an object, i.e. in the form:

```json
"objProp": {
  "name": "Alice",
  "age": 30
}
```

I don't think we ever cast to an instance of Map.

File: packages/client/src/components/app/forms/Form.svelte
Comment (melohagan) [reply packages/client/src/components/app/forms/Form.svelte:67]: Do you have an example of specific types being used in another svelte component?

File: packages/frontend-core/src/fetch/QueryArrayFetch.js
Comment (melohagan) [reply packages/frontend-core/src/fetch/QueryArrayFetch.js:10]: I don't think it's needed, but I didn't want to exclude it because I used the `QueryFetch` class as a template, and it had that check.

File: packages/frontend-core/src/utils/json.js
Comment (melohagan) [reply packages/frontend-core/src/utils/json.js]: Renamed to `hasSchema` to better indicate the purpose of the function.

**Review by samwho — APPROVED** [review]

File: packages/types/src/documents/app/row.ts
Comment (mike12345567) [inline packages/types/src/documents/app/row.ts:10]: One thing that we have been doing, in `packages/types/src/documents/app/table/schema.ts` - we've been defining the structure of Field schemas, the `QUERY_ARRAY` type - if it has any special properties etc it would be worth building out a specification for what the schema actually looks like/what is expected in the table/query schema definition. Right now these definitions are for tables, however someone could easily try to use this type for a table, so its probably best to define what is expected.

**Review by mike12345567 — APPROVED** [review]
LGTM - just one point about improving the typing around this.

**Comment by melohagan** [comment]
Holding off for the time being. Chatted with Michael, and ideally we want to make some typing improvements first: https://linear.app/budibase/issue/BUDI-7916/issue-with-how-the-schema-is-saved-between-rest-and-other-query-types

**Review by mike12345567 — APPROVED** [review]

**Pull request merged**
